lib/stores/S3Store.js

```diff
@@ -276,57 +276,81 @@ class S3Store extends DataStore {
-     */
-    _processUpload(metadata, req, current_part_number) {
-        return this._uploadPart(metadata, req, current_part_number)
-            .then(() => this.getOffset(metadata.file.id, true))
-            .then((current_offset) => {
-                if (parseInt(metadata.file.upload_length, 10) === current_offset.size) {
-                    return this._finishMultipartUpload(metadata, current_offset.parts)
-                        .then((location) => {
-                            log(`[${metadata.file.id}] finished uploading: ${location}`);
-
-                            this.emit(EVENTS.EVENT_UPLOAD_COMPLETE, {
-                                file: Object.assign({}, metadata.file, { location }),
-                            });
-
-                            this._clearCache(metadata.file.id);
-
-                            return current_offset.size;
-                        })
-                        .catch((err) => {
-                            throw err;
+   * uploads a stream to s3 using multiple parts
+   *
+   * @param {Object}                metadata            upload metadata
+   * @param {http<IncomingMessage>} req                 incoming request
+   * @param {Number}                currentPartNumber number of the current part/chunk
+   * @return {Promise<Number>}                         which resolves with the current offset
+   * @memberof S3Store
+   */
+    _processUpload(metadata, req, currentPartNumber) {
+        return new Promise((resolve, reject) => {
+            const splitter = new FileStreamSplitter(
+                {},
+                { maxChunkSize: this.part_size, directory: '.' }
+            );
+
+            const promises = [];
+
+            let pendingChunkFilepath = null;
+
+            stream.pipeline(req, splitter, (pipelineErr) => {
+                if (pipelineErr) {
+                    if (pendingChunkFilepath !== null) {
+                        fs.rm(pendingChunkFilepath, (fileRemoveErr) => {
+                            if (fileRemoveErr) {
+                                log(
+                                    `[${metadata.file.id}] failed to remove chunk ${pendingChunkFilepath}`
+                                );
+                            }
                         });
+                    }
                 }
 
-                return current_offset.size;
+                promises.push(pipelineErr ? Promise.reject(pipelineErr) : Promise.resolve());
+                resolve(promises);
+            });
+
+            splitter.on('chunkStarted', (filepath) => {
+                pendingChunkFilepath = filepath;
+            });
+
+            splitter.on('chunkFinished', (filepath) => {
+                pendingChunkFilepath = null;
+
+                const partNumber = currentPartNumber++;
+
+                const p = Promise.resolve()
+                    .then(() => {
+                        return this._uploadPart(metadata, fs.createReadStream(filepath), partNumber);
```

Comment:
Does _uploadPart take care of closing the read stream properly?

Reply:
the extended `_write` classes which is automatically called in the `pipeline` will close the stream. The underlying class, `Stream.Writeable`, also emits events whenever an error occurs. We also close the stream then:

```js
        this.on('error', this._finishChunk.bind(this));
```

Reply:
> the extended _write classes which is automatically called in the pipeline will close the stream

Ok, sounds good.